src/agent/rustjail/src/cgroups/fs/mod.rs

```diff
@@ -27,1120 +43,848 @@ macro_rules! sl {
+        cached_size: CachedSize::default(),
+    })
+}
 
-        let major = devno[0].parse::<u64>()?;
-        let minor = devno[1].parse::<u64>()?;
+/*
+examples(from runc, cgroup v1):
+
+    blkio.sectors
+    8:0 6792
+
+    blkio.io_service_bytes
+    8:0 Read 1282048
+    8:0 Write 2195456
+    8:0 Sync 2195456
+    8:0 Async 1282048
+    8:0 Total 3477504
+    Total 3477504
+
+    blkio.io_serviced
+    8:0 Read 124
+    8:0 Write 104
+    8:0 Sync 104
+    8:0 Async 124
+    8:0 Total 228
+    Total 228
+
+    blkio.io_queued
+    8:0 Read 0
+    8:0 Write 0
+    8:0 Sync 0
+    8:0 Async 0
+    8:0 Total 0
+    Total 0
+*/
+
+fn get_blkio_stat_blkiodata(blkiodata: &Vec<BlkIoData>) -> RepeatedField<BlkioStatsEntry> {
+    let mut m = RepeatedField::new();
+    if blkiodata.len() == 0 {
+        return m;
+    }
 
+    for d in blkiodata {
         m.push(BlkioStatsEntry {
-            major,
-            minor,
-            op,
-            value,
+            major: d.major as u64,
+            minor: d.minor as u64,
+            op: "".to_string(),
+            value: d.data,
             unknown_fields: UnknownFields::default(),
             cached_size: CachedSize::default(),
         });
     }
 
-    /*
-        if m.len() == 0 {
-            // return Err here? not sure about it
-            return Err(nix::Error::Sys(Errno:ENODATA).into());
-        }
-    */
-
-    Ok(m)
+    m
 }
-impl Blkio {
-    fn get_stats(&self, dir: &str) -> Result<BlkioStats> {
-        let mut m = BlkioStats::new();
-        let entry = get_blkio_stat(dir, BLKIO_IO_SERVICED)?;
-
-        if entry.len() == 0 {
-            // fall back to generic stats
-            // blkio.throttle.io_service_bytes,
-            // maybe io_service_bytes_recursive?
-            // stick to runc for now
-            m.io_service_bytes_recursive = get_blkio_stat(dir, BLKIO_THROTTLE_IO_SERVICE_BYTES)?;
-            m.io_serviced_recursive = get_blkio_stat(dir, BLKIO_THROTTLE_IO_SERVICED)?;
-        } else {
-            // cfq stats
-            m.sectors_recursive = get_blkio_stat(dir, BLKIO_SECTORS)?;
-            m.io_service_bytes_recursive = get_blkio_stat(dir, BLKIO_IO_SERVICE_BYTES)?;
-            m.io_serviced_recursive = get_blkio_stat(dir, BLKIO_IO_SERVICED)?;
-            m.io_queued_recursive = get_blkio_stat(dir, BLKIO_IO_QUEUED)?;
-            m.io_service_time_recursive = get_blkio_stat(dir, BLKIO_IO_SERVICE_TIME)?;
-            m.io_wait_time_recursive = get_blkio_stat(dir, BLKIO_IO_WAIT_TIME)?;
-            m.io_merged_recursive = get_blkio_stat(dir, BLKIO_IO_MERGED)?;
-            m.io_time_recursive = get_blkio_stat(dir, BLKIO_TIME)?;
-        }
 
-        Ok(m)
-    }
-}
+fn get_blkio_stat_ioservice(services: &Vec<IoService>) -> RepeatedField<BlkioStatsEntry> {
+    let mut m = RepeatedField::new();
 
-impl Subsystem for HugeTLB {
-    fn name(&self) -> String {
-        "hugetlb".to_string()
+    // do as runc
```

Comment:
Could you add a pointer to some code (e.g. a github source line)?

Reply:
This comment shouldn't be here, I' moved it.